cityofaustin/atd-moped
# Configure additional audit field triggers for component project subcomponents and refactor parent updates trigger

## Associated issues
https://github.com/cityofaustin/atd-data-tech/issues/16453

This PR:
- consolidates the functions called `update_parent_records_audit_logs` and `update_component_attributes_parent_records_audit_logs` into `update_audit_fields_with_dynamic_parent_table_name` which takes args to make this more reusable (thank you @johnclary for the [starter code](https://austininnovation.slack.com/archives/CNUEPKLB1/p1711550047469799) 🙏
- updates the triggers that used the two functions that were replaced with the new more general function
- adds triggers to update the audit fields of the `moped_proj_components_subcomponents` records and their parent component and project records

## Testing
**URL to test:** 

Local

**Steps to test:**
1. We need to test all tables that used the replaced functions which are:
   - `feature_drawn_lines`
   - `feature_drawn_points`
   - `feature_intersections`
   - `feature_signals`
   - `feature_street_segments`
   - `moped_proj_component_tags`
   - `moped_proj_component_work_types`
   - `moped_proj_components_subcomponents` - this one is new to this PR
2. To test, find a project and then create and update records in the following table. For the updates on all of these record types, the update can be tested by removing (soft-deleting) the features through the map editor and the tags, work types, and subcomponent through the component atrributes form.
3. On create, the parent project, parent component, and child record created should: 
   - have matching timestamps for `updated_at` that also matches the child record's `created_at`
   - have matching user IDs (your logged in ID) for `updated_by_user_id` that also matches the child record's `created_by_user_id`
4. On update, the parent project, parent component, and child record created should:
   - have matching timestamps for `updated_at`
   - have matching user IDs for `updated_by_user_id`

Query for all feature tables (i included all feature tables in one query but you can comment out joins and their matching selects unless you like sidescrolling 🐁, also h/t to @frankhereford for his features query in the original PR):
```sql
SELECT
	projects.project_id,
	components.component_id,
	projects.updated_at AS project_updated_at,
	projects.updated_by_user_id AS project_updated_by,
	components.updated_at AS component_updated_at,
	components.updated_by_user_id AS component_updated_by,
	fdl.updated_at,
	fdl.updated_by_user_id,
	fdl.created_at,
	fdl.created_by_user_id,
	fdp.updated_at,
	fdp.updated_by_user_id,
	fdp.created_at,
	fdp.created_by_user_id,
	fi.updated_at,
	fi.updated_by_user_id,
	fi.created_at,
	fi.created_by_user_id,
	fs.updated_at,
	fs.updated_by_user_id,
	fs.created_at,
	fs.created_by_user_id,
	fss.updated_at,
	fss.updated_by_user_id,
	fss.created_at,
	fss.created_by_user_id
FROM
	moped_project projects
	LEFT JOIN moped_proj_components components ON (components.project_id = projects.project_id)
	LEFT JOIN feature_drawn_lines fdl ON components.project_component_id = fdl.component_id
	LEFT JOIN feature_drawn_points fdp ON components.project_component_id = fdp.component_id
	LEFT JOIN feature_intersections fi ON components.project_component_id = fi.component_id
	LEFT JOIN feature_signals fs ON components.project_component_id = fs.component_id
	LEFT JOIN feature_street_segments fss ON components.project_component_id = fss.component_id
WHERE
	projects.project_id = <your project id>
```
Query for component tags:
```sql
-- Query to see component tags and associated project and project component record changes
SELECT
	mp.project_id,
	mp.updated_by_user_id,
	mp.updated_at,
	mpc.project_component_id,
	mpc.updated_at,
	mpc.updated_by_user_id,
	mpct.id,
	mpct.created_at,
	mpct.updated_at,
	mpct.created_by_user_id,
	mpct.updated_by_user_id,
	mpct.is_deleted
FROM
	moped_project mp
	LEFT JOIN moped_proj_components mpc ON mp.project_id = mpc.project_id
	LEFT JOIN moped_proj_component_tags mpct ON mpc.project_component_id = mpct.project_component_id
WHERE
	mp.project_id = <your project ID>;
```
Query for component work types:
```sql
SELECT
	mp.project_id,
	mp.updated_by_user_id,
	mp.updated_at,
	mpc.project_component_id,
	mpc.updated_at,
	mpc.updated_by_user_id,
	mpcwt.id,
	mpcwt.created_at,
	mpcwt.updated_at,
	mpcwt.created_by_user_id,
	mpcwt.updated_by_user_id,
	mpcwt.is_deleted
FROM
	moped_project mp
	LEFT JOIN moped_proj_components mpc ON mp.project_id = mpc.project_id
	LEFT JOIN moped_proj_component_work_types mpcwt ON mpc.project_component_id = mpcwt.project_component_id
WHERE
	mp.project_id = <your project ID>;
```

Query for component subcomponents:
```sql
SELECT
	mp.project_id,
	mp.updated_by_user_id,
	mp.updated_at,
	mpc.project_component_id,
	mpc.updated_at,
	mpc.updated_by_user_id,
	mpcs.component_subcomponent_id,
	mpcs.created_at,
	mpcs.updated_at,
	mpcs.created_by_user_id,
	mpcs.updated_by_user_id,
	mpcs.is_deleted
FROM
	moped_project mp
	LEFT JOIN moped_proj_components mpc ON mp.project_id = mpc.project_id
	LEFT JOIN moped_proj_components_subcomponents mpcs ON mpc.project_component_id = mpcs.project_component_id
WHERE
	mp.project_id = <your project id>;
```

---
#### Ship list
